idlparser/src/types/mod.rs

```diff
@@ -0,0 +1,374 @@
+    }
+
+    pub fn params(&self) -> &[FuncParam] {
+        &self.params
+    }
+
+    pub fn output(&self) -> &TypeDecl {
+        &self.output
+    }
+
+    pub fn is_query(&self) -> bool {
+        self.is_query
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct FuncParam {
+    name: String,
+    r#type: TypeDecl,
+}
+
+impl FuncParam {
+    pub(crate) fn new(name: String, r#type: TypeDecl) -> Self {
+        Self { name, r#type }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn r#type(&self) -> &TypeDecl {
+        &self.r#type
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Type {
+    name: String,
+    def: TypeDef,
+}
+
+impl Type {
+    pub(crate) fn new(name: String, def: TypeDef) -> Self {
+        Self { name, def }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn def(&self) -> &TypeDef {
+        &self.def
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TypeDecl {
+    Optional(Box<TypeDecl>),
+    Vector(Box<TypeDecl>),
+    Result {
+        ok: Box<TypeDecl>,
+        err: Box<TypeDecl>,
+    },
+    Id(TypeId),
+    Def(TypeDef),
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TypeId {
+    Primitive(PrimitiveType),
+    UserDefined(String),
+}
+
+#[derive(Debug, PartialEq)]
+#[repr(u8)]
+pub enum PrimitiveType {
+    Null,
+    Bool,
+    Char,
+    Str,
+    U8,
+    U16,
+    U32,
+    U64,
+    U128,
+    I8,
+    I16,
+    I32,
+    I64,
+    I128,
+}
+
+impl PrimitiveType {
+    pub(crate) fn str_to_enum(str: &str) -> Option<Self> {
+        match str {
+            "bool" => Some(PrimitiveType::Bool),
+            "char" => Some(PrimitiveType::Char),
+            "str" => Some(PrimitiveType::Str),
+            "u8" => Some(PrimitiveType::U8),
+            "u16" => Some(PrimitiveType::U16),
+            "u32" => Some(PrimitiveType::U32),
+            "u64" => Some(PrimitiveType::U64),
+            "u128" => Some(PrimitiveType::U128),
+            "i8" => Some(PrimitiveType::I8),
+            "i16" => Some(PrimitiveType::I16),
+            "i32" => Some(PrimitiveType::I32),
+            "i64" => Some(PrimitiveType::I64),
+            "i128" => Some(PrimitiveType::I128),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TypeDef {
+    Struct(StructDef),
+    Enum(EnumDef),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct StructDef {
+    fields: Vec<StructField>,
+}
+
+impl StructDef {
+    pub(crate) fn new(fields: Vec<StructField>) -> Self {
+        Self { fields }
+    }
+
+    pub fn fields(&self) -> &[StructField] {
+        &self.fields
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct StructField {
+    name: Option<String>,
+    r#type: TypeDecl,
+}
+
+impl StructField {
+    pub(crate) fn new(name: Option<String>, r#type: TypeDecl) -> Self {
+        Self { name, r#type }
+    }
+
+    pub fn name(&self) -> Option<&str> {
+        self.name.as_deref()
+    }
+
+    pub fn r#type(&self) -> &TypeDecl {
```

Comment:
I'd avoid such names. Maybe `kind` or another synonym?

Reply:
+1 for kind